openstreetmap-carto.lua

```diff
@@ -265,34 +509,75 @@ end
 
 --- Gets the roads table status for a set of tags
 -- @param tags OSM tags
--- @return 1 if it belongs in the roads table, 0 otherwise
+-- @return true if it belongs in the roads table, false otherwise
 function roads(tags)
     for k, v in pairs(tags) do
         if roads_info[k] and roads_info[k][v] and roads_info[k][v].roads then
             if not (k ~= 'railway' or tags.service) then
-                return 1
+                return true
             elseif not excluded_railway_service[tags.service] then
-                return 1
+                return true
             end
         end
     end
-    return 0
+    return false
+end
+
+--- Check if an object with given tags should be treated as polygon
+-- @param tags OSM tags
+-- @return true if area, false if linear
+function isarea (tags)
+    -- Treat objects tagged as area=yes polygon, other area as no
+    if tags["area"] then
+        return tags["area"] == "yes" and true or false
+    end
+
+   -- Search through object's tags
+    for k, v in pairs(tags) do
+        -- Check if it has a polygon key and not a linestring override, or a polygon k=v
+        for _, ptag in ipairs(polygon_keys) do
+            if k == ptag and v ~= "no" and not (linestring_values[k] and linestring_values[k][v]) then
+                return true
+            end
+        end
+
+        if (polygon_values[k] and polygon_values[k][v]) then
+            return true
+        end
+    end
+    return false
 end
 
---- Generic filtering of OSM tags
--- @param tags Raw OSM tags
--- @return Filtered OSM tags
-function filter_tags_generic(tags)
+--- Normalizes layer tags to integers
+-- @param v The layer tag value
+-- @return The input value if it is an integer between -100 and 100, or nil otherwise
+function layer (v)
+    return v and -- Check that layer is non-nil
+        string.find(v, "^-?%d+$") and tonumber(v) < 100 and tonumber(v) > -100 and -- Enforce numeric and range
+        v or nil -- return v, or nil if fails to match
```

Comment:
I am sorry but this is even more confusing now.

According to [the Lua manual](https://www.lua.org/manual/5.4/manual.html#3.4.8) `and` has higher priority than `or`.  Hence the line breaks are misleading.